plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -304,6 +316,94 @@ private static String quotedName(String name)
+            // never returns NULL for non-NULL input (unless it's the void transform),
+            // so NULL values and non-NULL values are always segregated.
+            ValueSet valueSet = domain.getValues();
+
+            boolean canEnforce = valueSet.getValuesProcessor().transform(
+                    ranges -> {
+                        ColumnTransform transform = PartitionTransforms.getColumnTransform(field, column.getType());
+                        MethodHandle targetTypeEqualOperator = typeOperators.getEqualOperator(
+                                transform.getType(), InvocationConvention.simpleConvention(FAIL_ON_NULL, NEVER_NULL, NEVER_NULL));
+                        for (Range range : ranges.getOrderedRanges()) {
+                            if (!canEnforceRangeWithPartitioningField(transform, range, targetTypeEqualOperator)) {
+                                return false;
+                            }
+                        }
+                        return true;
+                    },
+                    discreteValues -> false,
+                    allOrNone -> true);
+            if (canEnforce) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean canEnforceRangeWithPartitioningField(ColumnTransform transform, Range range, MethodHandle targetTypeEqualOperator)
+    {
+        io.trino.spi.type.Type type = range.getType();
+        if (!type.isOrderable()) {
+            return false;
+        }
+        if (!range.isLowUnbounded()) {
+            Object boundedValue = range.getLowBoundedValue();
+            Optional<Object> adjacentValue = range.isLowInclusive() ? getPreviousValue(type, boundedValue) : getNextValue(type, boundedValue);
+            if (adjacentValue.isEmpty() || yieldSamePartitioningValue(type, transform, boundedValue, adjacentValue.get(), targetTypeEqualOperator)) {
+                return false;
+            }
+        }
+        if (!range.isHighUnbounded()) {
+            Object boundedValue = range.getHighBoundedValue();
+            Optional<Object> adjacentValue = range.isHighInclusive() ? getNextValue(type, boundedValue) : getPreviousValue(type, boundedValue);
+            if (adjacentValue.isEmpty() || yieldSamePartitioningValue(type, transform, boundedValue, adjacentValue.get(), targetTypeEqualOperator)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean yieldSamePartitioningValue(
+            io.trino.spi.type.Type sourceType,
+            ColumnTransform transform,
+            Object first,
+            Object second,
+            MethodHandle targetTypeEqualOperator)
+    {
+        Object firstTransformed = transform.getValueTransform().apply(nativeValueToBlock(sourceType, first), 0);
+        Object secondTransformed = transform.getValueTransform().apply(nativeValueToBlock(sourceType, second), 0);
```

Comment:
add checks for results != null, to prevent some transform like void sneaking thru